libvast/test/type.cpp

```diff
@@ -0,0 +1,960 @@
+  CHECK(at);
+  CHECK_EQUAL(as_bytes(at), as_bytes(address_type{}));
+  CHECK(t != at);
+  CHECK(t < at);
+  CHECK(t <= at);
+  CHECK_EQUAL(fmt::format("{}", at), "addr");
+  CHECK_EQUAL(fmt::format("{}", address_type{}), "addr");
+  CHECK(!caf::holds_alternative<address_type>(t));
+  CHECK(caf::holds_alternative<address_type>(at));
+  const auto lat = type::from_legacy_type(legacy_address_type{});
+  CHECK(caf::holds_alternative<address_type>(lat));
+}
+
+TEST(subnet_type) {
+  static_assert(concrete_type<subnet_type>);
+  static_assert(basic_type<subnet_type>);
+  static_assert(!complex_type<subnet_type>);
+  const auto t = type{};
+  const auto st = type{subnet_type{}};
+  CHECK(st);
+  CHECK_EQUAL(as_bytes(st), as_bytes(subnet_type{}));
+  CHECK(t != st);
+  CHECK(t < st);
+  CHECK(t <= st);
+  CHECK_EQUAL(fmt::format("{}", st), "subnet");
+  CHECK_EQUAL(fmt::format("{}", subnet_type{}), "subnet");
+  CHECK(!caf::holds_alternative<subnet_type>(t));
+  CHECK(caf::holds_alternative<subnet_type>(st));
+  const auto lst = type::from_legacy_type(legacy_subnet_type{});
+  CHECK(caf::holds_alternative<subnet_type>(lst));
+}
+
+TEST(enumeration_type) {
+  static_assert(concrete_type<enumeration_type>);
+  static_assert(!basic_type<enumeration_type>);
+  static_assert(complex_type<enumeration_type>);
+  const auto t = type{};
+  const auto et = type{enumeration_type{{"first"}, {"third", 2}, {"fourth"}}};
+  CHECK(et);
+  CHECK(t != et);
+  CHECK(t < et);
+  CHECK(t <= et);
+  CHECK_EQUAL(fmt::format("{}", et), "enum {first: 0, third: 2, fourth: 3}");
+  CHECK(!caf::holds_alternative<enumeration_type>(t));
+  CHECK(caf::holds_alternative<enumeration_type>(et));
+  CHECK_EQUAL(caf::get<enumeration_type>(et).field(0), "first");
+  CHECK_EQUAL(caf::get<enumeration_type>(et).field(1), "");
+  CHECK_EQUAL(caf::get<enumeration_type>(et).field(2), "third");
+  CHECK_EQUAL(caf::get<enumeration_type>(et).field(3), "fourth");
+  CHECK_EQUAL(caf::get<enumeration_type>(et).resolve("first"), 0u);
+  CHECK_EQUAL(caf::get<enumeration_type>(et).resolve("second"), std::nullopt);
+  CHECK_EQUAL(caf::get<enumeration_type>(et).resolve("third"), 2u);
+  CHECK_EQUAL(caf::get<enumeration_type>(et).resolve("fourth"), 3u);
+  const auto let = type::from_legacy_type(
+    legacy_enumeration_type{{"first", "second", "third"}});
+  CHECK(caf::holds_alternative<enumeration_type>(let));
+  CHECK_EQUAL(caf::get<enumeration_type>(let).field(0), "first");
+  CHECK_EQUAL(caf::get<enumeration_type>(let).field(1), "second");
+  CHECK_EQUAL(caf::get<enumeration_type>(let).field(2), "third");
+  CHECK_EQUAL(caf::get<enumeration_type>(let).field(3), "");
+}
+
+TEST(list_type) {
+  static_assert(concrete_type<list_type>);
+  static_assert(!basic_type<list_type>);
+  static_assert(complex_type<list_type>);
+  const auto t = type{};
+  const auto lit = type{list_type{integer_type{}}};
+  CHECK(lit);
+  CHECK_EQUAL(as_bytes(lit), as_bytes(list_type{integer_type{}}));
```

Comment:
This line triggers the address sanitizer with GCC. Presumably because the destructor of the `list_type` is invoked immediately after `as_bytes` finishes. The span obviously can't extend the lifetime, so I think this should be expected.

I wonder if an rvalue overload `as_bytes(chunk&&) = delete;` would work as a warning system.

Reply:
Nice catch!

This occurs mostly because of how `CHECK_EQUAL` is implemented; it suffers from the same issues as the range-for loop as detailed in https://wg21.link/p2012.

It's very hard to find these kinda issues when macros are involved. We can generally forbid `as_bytes` on rvalues of all types and concrete types, which should help catch these kinda issues.

Reply:
Done. The deleted `as_bytes(list_type&&) = delete` actually found the line you commented on. I wonder why this didn't trigger the address sanitizer for me with clang.